### ISSUE TITLE
Error building ruby gem on Docker via Apple Silicon

### ISSUE THREAD

**Issue by bk-one** [grpc/grpc#26664]
### What version of gRPC and what language are you using?
Ruby 2.7 and latest gRPC (1.38 and latest master)

### What operating system (Linux, Windows,...) and version?
Linux via docker on Apple Silicon

### What runtime / compiler are you using (e.g. python version or version of gcc)
build error for ruby gem

### What did you do?
Using a minimalistic Docker File to build the ruby gem, using these files:

**Dockerfile**
``` 
FROM ruby:2.7

# throw errors if Gemfile has been modified since Gemfile.lock
RUN bundle config --global frozen 1

WORKDIR /usr/src/app

COPY Gemfile Gemfile.lock ./
RUN bundle install
```

**Gemfile** (tried both variants)
``` 
source 'https://rubygems.org'

gem 'grpc'
# gem 'grpc', github: 'grpc/grpc', submodules: true
```

**Gemfile.lock**
```
GEM
  remote: https://rubygems.org/
  specs:
    google-protobuf (3.17.3)
    googleapis-common-protos-types (1.1.0)
      google-protobuf (~> 3.14)
    grpc (1.38.0)
      google-protobuf (~> 3.15)
      googleapis-common-protos-types (~> 1.0)

PLATFORMS
  ruby

DEPENDENCIES
  grpc

BUNDLED WITH
   2.1.4
```

### What did you expect to see?
A successful build

### What did you see instead?

Long error log:
```
[ ... log shortened ... ]
#9 143.9 g++: fatal error: Killed signal terminated program cc1plus
#9 143.9 compilation terminated.
#9 143.9 make: *** [Makefile:823:
#9 143.9 /usr/local/bundle/gems/grpc-1.38.0/src/ruby/ext/grpc/objs/opt/src/core/ext/filters/client_channel/client_channel.o]
#9 143.9 Error 1
#9 143.9 make: *** Waiting for unfinished jobs....
#9 143.9 g++: fatal error: Killed signal terminated program cc1plus
#9 143.9 compilation terminated.
#9 143.9 make: *** [Makefile:823:
#9 143.9 /usr/local/bundle/gems/grpc-1.38.0/src/ruby/ext/grpc/objs/opt/src/core/ext/filters/client_channel/lb_policy/grpclb/grpclb.o]
#9 143.9 Error 1
#9 143.9 src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc:
#9 143.9 In member function ‘virtual
#9 143.9 grpc_core::RefCountedPtr<grpc_core::LoadBalancingPolicy::Config>
#9 143.9 grpc_core::{anonymous}::WeightedTargetLbFactory::ParseLoadBalancingConfig(const
#9 143.9 grpc_core::Json&, grpc_error**) const’:
#9 143.9 src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc:56:10:
#9 143.9 warning:
#9 143.9 ‘child_config.grpc_core::{anonymous}::WeightedTargetLbConfig::ChildConfig::weight’
#9 143.9 may be used uninitialized in this function [-Wmaybe-uninitialized]
#9 143.9    struct ChildConfig {
#9 143.9           ^~~~~~~~~~~
#9 143.9 src/core/ext/filters/client_channel/lb_policy/weighted_target/weighted_target.cc:657:45:
#9 143.9 note:
#9 143.9 ‘child_config.grpc_core::{anonymous}::WeightedTargetLbConfig::ChildConfig::weight’
#9 143.9 was declared here
#9 143.9          WeightedTargetLbConfig::ChildConfig child_config;
#9 143.9                                              ^~~~~~~~~~~~
#9 143.9 make: Leaving directory '/usr/local/bundle/gems/grpc-1.38.0'
#9 143.9 *** extconf.rb failed ***
#9 143.9 Could not create Makefile due to some reason, probably lack of necessary
#9 143.9 libraries and/or headers.  Check the mkmf.log file for more details.  You may
#9 143.9 need configuration options.
#9 143.9
#9 143.9 Provided configuration options:
#9 143.9 	--with-opt-dir
#9 143.9 	--without-opt-dir
#9 143.9 	--with-opt-include
#9 143.9 	--without-opt-include=${opt-dir}/include
#9 143.9 	--with-opt-lib
#9 143.9 	--without-opt-lib=${opt-dir}/lib
#9 143.9 	--with-make-prog
#9 143.9 	--without-make-prog
#9 143.9 	--srcdir=.
#9 143.9 	--curdir
#9 143.9 	--ruby=/usr/local/bin/$(RUBY_BASE_NAME)
#9 143.9
#9 143.9 extconf failed, exit code 1
#9 143.9
#9 143.9 Gem files will remain installed in /usr/local/bundle/gems/grpc-1.38.0 for
#9 143.9 inspection.
#9 143.9 Results logged to
#9 143.9 /usr/local/bundle/extensions/aarch64-linux/2.7.0/grpc-1.38.0/gem_make.out
#9 143.9
#9 143.9 An error occurred while installing grpc (1.38.0), and Bundler cannot continue.
#9 143.9 Make sure that `gem install grpc -v '1.38.0' --source 'https://rubygems.org/'`
#9 143.9 succeeds before bundling.
#9 143.9
#9 143.9 In Gemfile:
#9 143.9   grpc
```

**Comment by bk-one** [comment]
just FYI - running the same as `docker build --platform x86_64 -t my-app .` works fine.

**Comment by j-manu** [comment]
Please consider bumping the priority to P1. Since grpc is used by most of google cloud gems (via google-gax https://github.com/googleapis/google-cloud-ruby/search?q=google-gax&type=code), any project using google cloud ruby cannot be built natively on apple M1 macs using docker.

**Comment by paolo2299** [comment]
I'm not sure what's changed, but this is suddenly resolved for me (all the same steps as the OP, installs successfully with Docker For Mac Version 3.5.2 (3.5.2.18) and also with Version 3.6.0 (3.6.0.5487))

**Comment by bk-one** [comment]
can't confirm. It's still broken for me if I build from scratch without the explicit `platform linux/amd64` change.

**Comment by jtattermusch** [comment]
I'm a bit confused. Is this about building arm64 grpc linux gems or x86_64 linux grpc gems?

if this is about building arm64 linux gems, we do have an issue filed for that: https://github.com/grpc/grpc/issues/26391 (and last time I checked it was blocked on rake-compiler-dock not supporting arm64 darwin)

If this is about building x86_64 gems, that seems unnecessary since we already publish precompiled linux native gems with each grpc release (so I don't see why you would need to build your own, and why you would want to do that under docker on mac).

The "building under docker on mac" part of this issue seems irrelevant. Since docker on Mac is backed by a linux VM under the hood, the behavior should really be exactly the same as if you're trying to build under docker on a linux machine or simply trying to build the gem on a linux machine.

**Comment by bk-one** [comment]
@jtattermusch I do not want to build `x86_64` gems, no. I want to use whatever Docker decides is the best architectur for my Apple Silicon. I need to explicitly pass a platform to have the gem build, and I don't want to pass a platform.

**Comment by jtattermusch** [comment]
> @jtattermusch I do not want to build `x86_64` gems, no. I want to use whatever Docker decides is the best architecture for my Apple Silicon. I need to explicitly pass a platform to have the gem build, and I don't want to pass a platform.

ok, so you're basically saying you want to be able to build arm64 linux ruby gems. That's currently not supported, but it's tracked by https://github.com/grpc/grpc/issues/26391.  IIUC, Docker has nothing to do with this and this issue is basically a duplicate of that issue.

Closing this issue as dupe of https://github.com/grpc/grpc/issues/26391.

**Comment by hiroshi** [comment]
I'm not sure, but giving more memory (say 8GB instead of default 2GB) to Docker for mac may resolve some of build problem....

**Comment by tienle** [comment]
> I'm not sure, but giving more memory (say 8GB instead of default 2GB) to Docker for mac may resolve some of build problems....

Confirmed! Not sure why......

**Comment by Mayurifag** [comment]
Thanks @hiroshi for tip with memory allocation!

So basically, I've had a problem that previously working configuration stopped installing via bundler. Didn't know whats wrong, but grpc stopped to install, giving me HUGE >1MB log that something wrong.

Setting RAM >6GB helped. After installation in my case was totally fine to get RAM back. I hope, that will be fixed, that we need bigger RAM allocation on MacOS using Apple Silicon.


tags for search from google: In file included from third_party/boringssl-with-bazel grpc macos m1 issue

**Comment by Mayurifag** [comment]
actually problems still remains and im not sure why ticket is closed 😢

**Comment by ruvaleev** [comment]
I've passed this step with `--platform=linux/amd64`. amd64 - is to emulate x86 processors, so its not as good in performance terms, but at least it works.

**Comment by benbonnet** [comment]
same here. osx m1 catalina.
tried numerous combinations of ruby with specific versions of grpc.

1.45.0 is the only one building in a reasnoable amount of time on my machine.

1.5.x's just hangs forever.

Then, same machine, docker context : any versions hangs FOREVER. alpine, debian, various combo of ruby/grpc...

The most puzzling part is that this dockerfile builds relatively fine on ci or other laptops not running in osx m1

this gem is a fundamental for anything rubygoogle related; auch behaviours are really an issue. browsed all possible issues here, wirhout success
